### ISSUE TITLE
Would it be possible to have native/WebAssembly compilation with scriptc?

### ISSUE THREAD

**Issue by ChrisTrenkamp** [mathjax/MathJax#3604]
I'm using MathJax for server-side rendering equations with Java's GraaJS runtime.  GraalJS's slow warmup time has always been a source of frustration for me.  It's especially frustrating when we run our unit tests because it takes about 20 seconds to wait for GraalJS to warm up.  I'm willing to trade a slightly slower math renderer for instantaneous startup times.

Yesterday, I heard about scriptc, which claims to compile Typescript to native binaries and WebAssembly: https://github.com/vercel-labs/scriptc .  This seems to be a perfect fit.  I would love to replace our GraalVM interpreter with a lightweight WebAssembly or native binary.  It's a young project, however, so bugs and slow binaries should be expected.

There's other projects like AssemblyScript, but it doesn't support unions, which MathJax heavily relies on.  MathJax also implicitly depends on the DOM API, which scriptc does not supply, so you get errors like the following:

```
node_modules/.pnpm/@mathjax+src@4.1.3/node_modules/@mathjax/src/mjs/mathjax.d.ts:1:1 - error SC0001: Cannot find name 'Document'.
node_modules/.pnpm/@mathjax+src@4.1.3/node_modules/@mathjax/src/mjs/mathjax.d.ts:1:1 - error SC0001: Cannot find name 'Window'.
node_modules/.pnpm/@mathjax+src@4.1.3/node_modules/@mathjax/src/mjs/adaptors/HTMLAdaptor.d.ts:1:1 - error SC0001: Cannot find name 'Worker'.
.pnpm/@mathjax+src@4.1.3/node_modules/@mathjax/src/mjs/core/Tree/WrapperFactory.d.ts:1:1 - error SC0001: Type 'WC' does not satisfy the constraint 'FactoryNodeClass<WW>'.
  Type 'WrapperClass<N, C, WW>' is not assignable to type 'FactoryNodeClass<WW>'.
    Types of parameters 'factory' and 'factory' are incompatible.
      Property 'wrap' is missing in type 'Factory<WW, FactoryNodeClass<WW>>' but required in type 'WrapperFactory<N, C, WW, WrapperClass<N, C, WW>>'.
```

Is this worth investigating?  I think making MathJax available on non-JS runtimes would be a gamechanger.

**Comment by dpvc** [comment]
Sorry to leave you hanging for so long.  I have not had the chance to look into this more carefully, but I suspect that the first three issues can be resolved by adding the `DOM` library to the `lib` array in `tsconfig.json` file for your project.

I'm not sure what to say about the last one, as the code compiles in our repository, so I don't know why the `.d.ts` file would cause a problem.  That will take some investigation, I'm afraid.

I have not ever worked with WebAssembly myself, so I don't know whether it makes sense for MathJax as a whole.  For example, I'm not sure how it would deal with the dynamic loading of components perhaps that would work just fine, but I don't know.  I suspect that self-contained applications that use direct loading of Mathjax modules could be made to work.

It is probably worth a closer look.
